### PR TITLE
implement SCL3 httpd redirects in Django

### PR DESCRIPTION
**Pull request by metadave**
previously: https://github.com/mozilla/kuma/pull/4215

This PR uses @pmac's https://github.com/pmac/django-redirect-urls lib to implement SCL3 Apache httpd redirects in Python/Django. The complex rewrites retain original mod_rewrite rules as comments, while the remaining are uncommented simple static redirects. Test layout is similar to the [Bedrock redirect tests](https://github.com/mozilla/bedrock/tree/master/tests/redirects). 

The only rewrite rule that isn't implemented in this PR is as follows:

```
RewriteCond %{SERVER_NAME} ^developer\.mozilla\.com$
RewriteRule (.*) http://developer.mozilla.org$1 [R=301,L]
```

which redirects requests to `developer.mozilla.com/*` --> `developer.mozilla.org/*`

Tracked in [this infra issue](https://github.com/mozmar/infra/issues/180)

### CODE REVIEW

File: tests/redirects/test_redirects.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+from .map_301 import GITHUB_IO_URLS
+from .map_301 import MOZILLADEMOS_URLS
+
+# while these test methods are similar, they're each testing a
+# subset of redirects, and it was easier to work with them separately.
+@pytest.mark.headless
+@pytest.mark.nondestructive
+@pytest.mark.parametrize('url', REDIRECT_URLS)
+def test_redirects(url):
+    url['base_url'] = "https://developer.mozilla.org"
```
Comment (jwhitlock) [inline tests/redirects/test_redirects.py]: This should match [bedrock](https://github.com/mozilla/bedrock/blob/master/tests/redirects/test_redirects.py#L18-L19), using the ``base_url`` py.test fixture:

```
def test_redirects(url, base_url):
    url['base_url'] = base_url
```

You can then test it against staging:

```
py.test tests/redirects -vv --base-url https://developer.allizom.org --pdb
```
or the local Docker:

```
py.test tests/redirects -vv --base-url http://localhost:8000 --pdb
```

The ``--pdb`` drops you into the [Python debugger](https://docs.python.org/2/library/pdb.html), where you can print out the expected and actual values ("q" and Enter to quit).  If we were on py.test 3.x, the assets may be more useful (see PR #4210), but w/ py.test 2.x, you have to do your own test debugging.

File: tests/redirects/map_301.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+    url_test("/media/redesign/js.foo", "/static/js.foo"),
+    url_test("/media/js.foo", "/static/js.foo"),
+
+    url_test("/media/redesign/fonts.foo", "/static/fonts.foo"),
+    url_test("/media/fonts.foo", "/static/fonts.foo"),
+
+    url_test("/media/uploads/demos/foobar123",
+             "/docs/Web/Demos_of_open_web_technologies/",
+             status_code=requests.codes.found),
```
Comment (jwhitlock) [inline tests/redirects/map_301.py]: This fails against ``--base_url=localhost:8000``.  It expects a 302 Found, but gets a 301 Moved Permanently.

File: tests/redirects/map_301.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+    url_test("/foo//bar", "/foo_bar"),
```
Comment (jwhitlock) [inline tests/redirects/map_301.py]: This fails against ``--base_url=localhost:8000``, getting ``http://localhost:8000/foo//foo_bar``.

It also fails against ``--base_url=https://developer.allizom.org``, getting ``https://en-US/foo/bar``.  I think the staging server's Apache config is wrong.

I also question the purpose of this pattern, but I guess it was a problem at some point (see [bug 779501](https://bugzilla.mozilla.org/show_bug.cgi?id=779501))

File: tests/redirects/map_301.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    url_test("/foo//bar//baz", "/foo_bar_baz"),
```
Comment (jwhitlock) [inline tests/redirects/map_301.py]: This fails against ``--base_url=localhost:8000``, getting ``http://localhost:8000/bar_baz``

It also fails against ``--base_url=https://developer.allizom.org``, getting ``https://developer.allizom.org/f/en-US/foo/bar/baz``.

WEIRD.

**Comment by jwhitlock** [comment]
There's some missing stuff:

* 3 Tests that pass in developer.mozilla.org but fail in developer.allizom.org or Docker dev
* The [.htaccess](https://github.com/mozilla/kuma/blob/master/configs/htaccess) redirects
* Tests are not integrated into acceptance tests

but, I'm happy with what is here so far. It's up to @metadave if he wants to keep working in this PR, or merge and add further refinements in a future PR.

**Comment by metadave** [comment]
@jwhitlock I pushed some fixes to get all the tests to pass on `http://localhost:8000` and `https://developer.mozilla.org`, however there are 2 remaining failures on `https://developer.allizom.org` related to the following tests:

```
    url_test("/docs/Mozilla/Projects/NSPR/Reference/I//O_Functions", "/docs/Mozilla/Projects/NSPR/Reference/I_O_Functions"),
    url_test("/docs/Mozilla/Projects/NSPR/Reference/I//O//Functions", "/docs/Mozilla/Projects/NSPR/Reference/I_O_Functions"),
```

which correspond to the following rewrites:

```
RewriteRule ^(.*)//(.*)//(.*)$ $1_$2_$3 [R=301,L,NC]
RewriteRule ^(.*)//(.*)$ $1_$2 [R=301,L,NC]
```

I'd prefer to submit * the [.htaccess](https://github.com/mozilla/kuma/blob/master/configs/htaccess) redirects in a separate PR.

**Comment by jwhitlock** [comment]
Nice work @metadave. I'll diff the staging and production configuration, and see why there is a difference for those URL patterns.

**Comment by jwhitlock** [comment]
I've looked at the staging and production Apache configuration, and the two given rewrite rules are missing from staging.  I've opened [infra bug 1365774](https://bugzilla.mozilla.org/show_bug.cgi?id=1365774) to request syncing the two configurations.

**Comment by jwhitlock** [comment]
The configurations have been synced, and the tests now pass against staging: ``py.test tests/redirects/``.  I think we're ready to add these tests to the integration test suite.

**Pull request merged**